Modal work.

bnhovde opened the pull request:
Had 30 minutes to kill.
- Added escape key functionality.
- Made the overlay clickable to close

I've also added some data-attributes to the body. I've found this really useful in other projects where I want to quickly see what JS components a page has by just looking at the body attributes.

As always, happy to review/change any approach here.

bnhovde committed:
Modal work. Added escape button listener, added overlay click-to-close.

chris-pearce commented:
LGTM.

> I've also added some data-attributes to the body. I've found this really useful in other projects where I want to quickly see what JS components a page has by just looking at the body attributes.

That's a good idea, I like it :+1:.

I was thinking—and this isn't a TODO for this PR as we're all working on this component (well @bnhovde is :smile:)—that perhaps this component should be called **Dialog** as from what I understand **Modal** is a type of dialog behaviour, from MDN:

> Dialogs can be modal or non-modal. When a modal dialog appears on the screen, it's not possible to interact with any page content outside the dialog. In other words, the main application UI or page content is considered to be temporarily disabled as long as the modal dialog is showing. For non-modal dialogs it is still possible to interact with content outside of the dialog while the dialog is showing. Note that for non-modal dialogs there will have to be a global keyboard shortcut that allows focus to be moved between opened dialogs and the main page. For more information see the Managing Modal and Non Modal Dialogs guide. 
> —https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_dialog_role

So we could have a new addition to the `userOptions` object: `isModal` that'll disable all the ways to close a dialog except when an action within the dialog is actioned e.g. 2 buttons: **OK** and **Cancel**?

There is also an **Alert Dialog** type: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alertdialog_role which are always **modal**.

Out of interest I thought I'd see where the `dialog` element support is at, it's dismal: http://caniuse.com/#search=dialog%20element. More info from MDN: https://developer.mozilla.org/en/docs/Web/HTML/Element/dialog.

bnhovde commented:
Yeah Dialog makes sense then. I've always thought modal was a term for overlays, but looking at what you've found here it's not and **Dialog** is the right name for the component.

RE: Like the idea of the `isModal` (or maybe `dialogType` to encompass the alert variant and others) config option. This should be a quick job. I'll have a go at that tomorrow.

Never knew there was a `dialog` element. Maybe support is so poor because people don't know it exists, create their own solutions and don't demand it implemented.

bnhovde committed:
Renamed modal to dialog and implemented new functionality. Cleaned up dead js files. Started implementing testing.

bnhovde commented:
Finally got around to this.
- Renamed all references of modal to dialog
- Added optional `isModal` config option that determines if escape closes modal or not
- Updated demo
- Started writing tests for the module but hit a wall. Will pick this up later.

Other stuff:
- Deleted dead .js files (library etc)

chris-pearce commented on src/components/dialog/dialog.js:
```diff
@@ -0,0 +1,100 @@
+        closeBtn:       '.js-dialog-close-btn',
+        backdrop:       '.js-dialog-backdrop',
+        isModal:        false,
+        showBackdrop:   true
```
One thing me and @davislurve  are doing at work is starting all `boolean` variables with _is_ or _has_ so this one would be something like `hasBackdrop`. How do you feel about that naming convention?

bnhovde replied:
Yeah like it. We need a style guide don't we? Hehe..

chris-pearce commented on src/components/dialog/dialog.js:
```diff
@@ -0,0 +1,100 @@
+export default function UIDialog(userOptions) {
+
+    var DOM,
+        state = {};
```
Just curious why this wouldn't be a `let`?

bnhovde replied:
Old habits mate! Good spot

chris-pearce commented on src/components/dialog/dialog.js:
```diff
@@ -0,0 +1,100 @@
+            'backdrop': document.querySelectorAll(settings.backdrop)[0]
+        };
+        
+        // Check if gallery exists, return if not
```
Replace _gallery_ with _dialog_ :+1:.

chris-pearce commented on src/components/dialog/dialog.js:
```diff
@@ -0,0 +1,100 @@
+        };
+        
+        // Check if gallery exists, return if not
+        if (DOM.dialog === undefined) { return false; }
```
How do we feel about 1 line blocks? Personally I'm not a fan but happy to go with what the majority wants.

bnhovde replied:
I'll also go with the majority.

My personal preference is towards single-line for single statements, for both js and css.

For example, when a map function only performs one operation:

`[1, 2, 3].map(function (num) { return num * 2 })`

but with a second operation comes wrappage:

```
[1, 2, 3, 4].map(num => {
    let multiplier = 2 + num
    return num * multiplier
})
```

Anyway, this is one of those cases where there's no real rational behind it other than I think it looks nicer, so let democracy decide!

Another one for the style guide.

chris-pearce commented on src/components/dialog/dialog.js:
```diff
@@ -0,0 +1,100 @@
+    /**
+    * @function init
+    * @desc Initialises the dialog
+    */ 
```
Indent the `*`s by 1 space :+1:.

chris-pearce commented on src/components/dialog/dialog.js:
```diff
@@ -0,0 +1,100 @@
+    
+    /**
+     * @function keyHandler
+     * @desc Checks to see if escape (key 27) has been pressed and dialog not modal
```
Do we need to the _and dialog not modal_ check on `closeBtn` and `backdrop` to?

chris-pearce commented:
Looking good @bnhovde :+1:. There's more functionality around accessibility but not sure if you plan to get to that later? If you need a hand with any of it I can provide you with the specs. Cheers.

bnhovde commented:
@chris-pearce Yeah I think It'll need a round in the accessibility ring!

I'm just merging this in now, forgot it was still pending.
